exchanges/gateio/gateio_websocket_request_spot.go

```diff
@@ -0,0 +1,278 @@
+package gateio
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/request"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream"
+)
+
+var (
+	errBatchSliceEmpty  = errors.New("batch cannot be empty")
+	errNoOrdersToCancel = errors.New("no orders to cancel")
+	errEdgeCaseIssue    = errors.New("edge case issue")
+	errChannelEmpty     = errors.New("channel cannot be empty")
+)
+
+// websocketLogin authenticates the websocket connection
+func (g *Gateio) websocketLogin(ctx context.Context, conn stream.Connection, channel string) (*WebsocketLoginResponse, error) {
```

Comment:
> I still need declare an anon func on the authenticate field so it will be something like this and in future the futures endpoints in the PR above this one for its connection:

Okay, ignore that, then. If you'd need an anonymous function and you're *not* injecting state into it, then what you've done is sanest.

> Also I don't know if we ever need the return value (APIKEY or Connection ID) but it gives me a clue that there is something else returned so I would rather for the moment keep it for potential future usage. 🤷

Nope, I still don't agree. Anything that's coming out of that API is something that websocketLogin should be handling.
You have no consumer for this (even conceivably, I think), so there's nobody else to process that response but the function itself. Any state it gets, it needs to find a place to store (WS or exch I guess).
But if you don't have a consumer for the information, don't return it 🙂 
TL;DR: Don't be a hoarder 😂 

> With regard to the dropping prefix that is understandable maybe we should have that as its own PR across exchanges for a standardisation pass? I have addressed the inconsistency though and will push that up soon.

Agreed - Out of scope and would be wrong to do it here.